utils/docs/lepton-xyrs.1.in
.\"Created with GNOME Manpages Editor Wizard
.\"http://sourceforge.net/projects/gmanedit2
.TH \fBlepton-xyrs\fP 1 "@DATE@" "Lepton EDA" "@VERSION@"

.SH NAME
\fBlepton-xyrs\fP \- is a program to batch process XYRS files.

.SH SYNOPSIS
.B \fBlepton-xyrs\fP 
.RI \fBfile\fP
.RI [OPTIONS] 
.br

.SH DESCRIPTION
XYRS files are usually generated by PCB design programs, and are used by board assemblers.
.PP
Format of XYRS files is discussed in FILES section.
.PP
Processing of these files is often needed by board assemblers or designers to suit their 
process and tools flows.
.PP
\fBlepton-xyrs\fP helps batch-processing these files. Rules and commands can be defined to make 
batch-processing easy.
.PP
\fBlepton-xyrs\fP consists on a perl library providing the core functions, and a perl script providing some basic functionality.
.PP
It takes some arguments from the command line: input and output files, some modifiers, and an action-file or an action-string.
.PP
The action-file or action-script contains some commands. For every line of the input file, \fBlepton-xyrs\fP executes those commands, which can modify the contents of the line, 
and then writes the resulting line to the output file.
.PP
Since \fBlepton-xyrs\fP is written in perl, actions are also written in this language.

.SH OPTIONS
.B
.IP       --help 
Display  a  usage  message  on standard output and exit successfully.
.BI      \--verbose 
Display more information messages when processing files.
.TP
.BI      \--process-comments 
By default, lepton-xyrs doesn't process lines starting with '*' and '#' characters.

It just copies these lines to the output, without further processing.

With this option, it processes these lines as well.
.TP
.BI      \--tabulate 
Give each field in the output file a fixed length, enough to contain the longest string.
		Using this option make the output file easier to read by humans.
.TP
.BI      \--caseinsensitive 
Ignore case distinctions when comparing patterns.
.TP
.BI      \--adjust "\| FILE\ " 
File with the commands to process the input file.
.TP
.BI      \--eval "\| PATTERN\ " 
Interpret PATTERN as the commands to process the input file.
.TP
.BI      \--output "\| FILE\ " 
Save the output to the specified FILE. 

If FILE string is "-" (without double quotes), then output is redirected to standard output STDOUT.
.TP
.BI      \--output-delimiter "\| CHAR\ "  
Use the given CHAR as an output field delimiter.

.SH "PREDEFINED VARIABLES"
.PP
lepton-xyrs defines automatically the following variables:
.PP
.TP
.B REF_COL
column number (starting at 0) where the component reference is.
.TP
.B FOOTPRINT_COL
column number (starting at 0) where the component footprint is.
.TP
.B X_COL
column number (starting at 0) where the component's X location coordinate is.
.TP
.B Y_COL
column number (starting at 0) where the component's Y location coordinate is.
.TP
.B ANGLE_COL
column number (starting at 0) where the component rotation angle is.
.TP
.B LAYER_COL
column number (starting at 0) where the component's layer side location is.
.TP
.B VALUE_COL
column number (starting at 0) where the component's value is. Usually this is the component's part number.
.TP
.B LINE_NUMBER
this is the line number of the input file that is being currently processed.
.TP
.B LINE
array where each array element is a field or column of the input file's line being processed.
.SH "BUILTIN COMMANDS"
All the following commands accept a \fIchecklist\fP parameter.
A \fIchecklist\fP is a list of one or more \fIcheck_items\fP.
A \fIcheckitem\fP is a list of two single elements:
.TP
.B column_number
is the data index number 'n' of the global variable \fBLINE\fP to be checked.
The first element of the data has the number 1.
.TP
.B pattern 
is the regular expression to be matched.
.PP
The commands will only be executed on those lines that match all the \fIcheckitems\fP in the \fIchecklist\fP.

.TP
.B \fBdel_line\fP \fIchecklist\fP;
Delete the line (global variable \fBLINE\fP) if \fIchecklist\fP is matched.

Returns \-1 if error, 0 if not match, 1 if match and changed.
.TP
.B rotate_comp \fIangle_col\fP, \fIangle\fP, \fIchecklist\fP;
rotate the component the degrees given in \fIangle\fP parameter, if \fIchecklist\fP is matched. 

The angle column number is \fIangle_col\fP.

Returns 1 if match and changed, 0 if not match, \-1 if error.
.TP
.B subst_col_val \fIcol\fP, \fIvalue\fP, \fIchecklist\fP;
Replace a column value by the new value \fIvalue\fP, if \fIchecklist\fP is matched.

Returns 1 if match and changed, 0 if not match, \-1 if error.

.TP
.B change_col_units \fIunits\fP, \fIcolumn_numbers\fP;
Change units of a given column number. More than one column number can be specified.

\fIUnits\fP is a string with the desired units. Only "mm" (milimeters), "in" (inches) and "mil" (thousands of an inch) are supported.

It is required that numbers to be converted are followed their units (see supported units above). Otherwise, the number is not changed.

.TP
.B add_number_to_col \fIcol_number\fP, \fIvalue\fP, \fIchecklist\fP;
Adds a number to the value in a given column number.
Note: offset and the value to be changed can be in different units.

Returns \-1 if error, \-2 if warning, and 1 if success.

.TP
.B translate_col_val \fIcol_number\fP, \fIstring\fP, \fIsubstitution\fP, \fIchecklist\fP;
Translate a \fIstring\fP in the column \fIcol_number\fP if \fIchecklist\fP is matched.

\fISubstitution\fP is an expression with the new string. Old column value can be used here.

Returns \-1 if error, 0 if not match, 1 if match and changed.

Example:
  translate 2, '^([0-9]+)n$','sprintf("%dnF",$1)', 3, 'C[0-9]+';
    if the value in column 3 is C followed by a number, then 
    if the value in column 2 is a number followed by 'n', translate it to the same value followed by 'nF'.
    If there is 'C10' in column 3 and '10n' in column 2, 
    change '10n' to '10nF'.

.TP
.B mul_col_val \fIcol_number\fP, \fIfactor\fP, \fIchecklist\fP;
Multiply the number in the specified column number \fIcol_number\fP by the given \fIfactor\fP, if \fIchecklist\fP is matched.

The number in the specified column number may have units at the end.

Returns \-1 if error, 0 if not match, 1 if match and changed.

.TP
.B swap_columns \fIcol_number1\fP, \fIcol_number2\fP, \fIchecklist\fP;
Swap columns \fIcol_number1\fP and \fIcol_number2\fP if \fIchecklist\fP is matched.

Returns \-1 if error, 0 if not match, 1 if match and changed.

.TP
.B insert_column \fIcol_number\fP, \fInew_column_value\fP;
Insert a new column in the given \fIcol_number\fP position (0 if it's going to be the first column), with the value \fInew_column_value\fP.

Returns \-1 if error, 1 if the new column was inserted.

.SH "RETURN VALUE"
The return value of \fBlepton-xyrs\fP is an integer with the following possible values:
.TP
	1 if the command found a match and the execution was successful
.PP
	0 if there was no match
.PP
	\-1 if there was an error
.PP
	\-2 if there was a warning
.PP

.SH FILES
.TP
.B       XYRS files
The format of these files is not standard, and depends on the application used to generate it.

The data stored in these files is usually a header, and one line per component of the board.

For each component, at least the following information is needed by assemblers:

	- Reference

	- Coordinates (X and Y)

	- Rotation angle

	- Side where the component is located (top or bottom)

There can be more information, such as description, footprint, but this depends on the 
design program an its configuration.

This information is organized in lines, one line per component, and each line is divided in as much fields or columns as needed.
Field delimiters are used between columns, but field delimiters are not standardized and they depend on the program used to generate the XYRS file.

The program that generates the XYRS file usually write a header  (usually the first two or thre lines in the file) with a column title.

\fBlepton-xyrs\fP tries to guess the file format, and column numbers, based on this header. Thus, the known column titles are the following:

	- for reference designator: \fIDesignator\fP, \fIRefDesignator\fP, \fIRefDes\fP.

	- for footprint: \fIFootprint\fP, \fITopCell\fP, \fIDescription\fP (only if there is no other "TopCell" column title).

	- for X coordinate: \fIMid X\fP, \fIX\fP.

	- for Y coordinate: \fIMid Y\fP, \fIY\fP.

	- for rotation angle: \fIRotation\fP, \fIrotation\fP, \fIRot\fP.

	- for location side: \fITB\fP, \fISide\fP, \fItop/bottom\fP.

	- for value: \fIComment\fP, \fIPartNumber\fP, \fIValue\fP.

.TP
.B       Action files
An action file contains a list of commands, using perl language, that can include any call to a built-in function or use any variable defined by \fIlepton-xyrs\fP.

Each action file should end with the following line (with a carrier return at the end of the line):

	1;

.SH EXAMPLES
.TP
.B Run an action file:
Run the action file \fIyour_comands.txt\fP with the XYRS file \fIyour_xyrs_file.txt\fP and write the output result to the file \fIoutput_file.txt\fP:

	lepton-xyrs  your_xyrs_file.txt \-\-adjust your_commands.txt \-\-output output_file.txt
.PP
.TP
.B Run a command specified in the command line:
Run the commands \fIyour_commands\fP wiith the XYRS file \fIyour_xyrs_file.txt\fP and write the output result to the file \fIoutput_file.txt\fP:

	lepton-xyrs  your_xyrs_file.txt \-\-eval "your_commands" \-\-output output_file.txt
.PP

.TP
.B Delete some lines matching a pattern. 
Parse the XYRS file \fIyour_xyrs_file.txt\fP, delete all lines having a R followed by a number in column number 2, and write the output result to the file \fIoutput_file.txt\fP:

	lepton-xyrs your_xyrs_file.txt \-\-eval "del 2, 'R[0\-9]+';" \-\-output output_file.txt
.PP
.TP
.B Delete some lines matching a pattern, using the column number guessed by \fBlepton-xyrs\fP. 
Parse the XYRS file \fIyour_xyrs_file.txt\fP, delete all lines having a R followed by a number in reference designator column, and write the output result to the file \fIoutput_file.txt\fP:

	lepton-xyrs your_xyrs_file.txt \-\-eval "del \\$REF_COL, 'R[0\-9]+';" \-\-output output_file.txt
.PP
.TP
.B Rotate some lines matching a pattern.
Parse the XYRS file \fIyour_xyrs_file.txt\fP, rotate by 90º all lines having a R followed by a number in reference designator column, and write the output result to the file \fIoutput_file.txt\fP. Use the rotation  and reference column number guessed by \fIlepton-xyrs\fP:

	lepton-xyrs your_xyrs_file.txt \-\-eval "rotate \\$ANGLE_COL, 90, \\$REF_COL, 'R[0\-9]+';" \-\-output output_file.txt
.PP
.TP
.B Change all numbers to 'mm' units.
Parse the XYRS file \fIyour_xyrs_file.txt\fP, convert all numbers in column 3 and 5 to its equivalent in mm, and write the output result to the file \fIoutput_file.txt\fP:

	lepton-xyrs your_xyrs_file.txt \-\-eval "change_units 'mm', 3, 5;" \-\-output output_file.txt
.PP
.TP
.B Replace a value matching a pattern with another value.

NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2016 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf 1.9.3
====================================

* The gEDA/gaf source tarball no longer includes libintl.  When
  building gEDA/gaf with internationalisation support, a GNU
  gettext-compatible libintl is now required.

* gEDA/gaf core tools will now automatically rotate per-user log
  files.  250 log files per tool will be retained from the last 24
  hours, and 50 log files per tool from all time.

* gEDA/gaf tools now search for data and configuration files in the
  standard XDG directories.  By default, these are:

  - `/usr/local/share/gEDA/` and `/usr/share/gEDA` for system data
  - `/etc/xdg/gEDA/` for system configuration
  - `$HOME/.local/share/gEDA/` for per-user data
  - `$HOME/.config/gEDA/` for per-user configuration.

  However, gEDA/gaf tools will still use `$HOME/.gEDA` for per-user
  files if that directory exists, they still obey the `$GEDADATA` and
  `$GEDADATARC` environment variables, and by default the search path
  includes the installation prefix selected when compiling gEDA/gaf.

* The configure script now supports an `--enable-relocatable` switch.
  When specified, gEDA/gaf tools do not have the configured prefix
  compiled into them.

* The configure script now supports a `--disable-deprecated` switch.
  When specified, various deprecated features and behaviour are
  disabled, which may break backward-compatibility with some
  configurations.

* A new Scheme API function, `log!`, has been added to the `(geda
  log)` module.  It allows Scheme code to emit log messages in the
  same way that the tools' C code does.

* Gschem multiattrib dialog improvements:

  - Shift+Tab can be used to move focus backward from `value' to
    `name'.

  - Tooltip to the `value' text entry has been added explaning how
    to enter newline or tab.

  - Horizontal scrollbar for `value' has been added to prevent
    resizing of the dialog.

* Gschem undo information now contains center of viewport and
  current page scale, so if gschem window size changes before
  undo, page will be zoomed accordingly.

* Faster redrawing of pages in gschem.

* Menu entry to launch Scheme REPL in gschem's background terminal
  has been added.

* Changed calculation of the size of bus and net junctions.

* `Bom` gnetlist backend has been refactored, all its functions
  have help now.

* View bounds in gschem window has been changed so symbols can be
  edited without translation.

* New Scheme module for part list creation has been added for
  gnetlist.

* Netlist output of several gnetlist backends is sorted by package
  names.

* New Scheme procedures for working with source library have been
  added to the `(geda library)` module: source-library-contents
  and set-source-library-contents!.  Please see the "geda-scheme"
  info manual for more details.


Notable changes in gEDA/gaf 1.9.2
=================================
* Build system changes:

  - Guile 2.0 or newer is now required.

  - top-level control file allowing continuous integration using
    Travis CI has been added.

* Nullor expansion has been fixed in the `spice' and `spice-sdb'
  backends.

  - Element name prefix "E_" has been substituted for "E-" to prevent
    errors output by some simulators (e.g. ngspice).

  - A default gain value has been added in the nullor expansion code.

* Two obsolete scripts verilog2vhdl.sh and script.sed have been
  removed.

* Undocumented and ambiguous feature of the `spice-sdb' backend for
  gnetlist consisting in the possibility of using the "value="
  attribute instead of "file=" only for include blocks and only in
  embedding mode has been removed.

* The following scripts and programs have been moved from the `utils'
  directory to a new `contrib' directory and are no longer installed
  by default:

         smash_megafile          olib
         gmk_sym                 convert_sym
         sarlacc_schem           sarlacc_sym
         pads_backannotate       sw2asc
         gnet_hier_verilog.sh

* `gpstoimage' has been removed from the distribution; equivalent
  functionality is provided by the `gaf' command.

* `convert_sym.awk' has been removed from the distribution; equivalent
  functionality is provided by the `convert_sym' command.

* gnetlist now supports output of netlists to the standard output.

* New Scheme procedures and hooks have been added.  Please see the
  "geda-scheme" info manual for more details.

  - `set-component-with-transform!' is a modified version of
    `set-component!', which applies all the specified transformations
    to a component and immediately updates it.

  - `auto-refdes-reset!' resets components' refdeses

  - `copy-objects-hook' is called whenever an object is copied, it can
    be used, for example, to reset component's refdes on copy.

* The following gschem configuration items have been added:

  - `third-button-cancel' cancels draw actions by the third mouse
    button in mousepan mode.

  - `grid-mode' sets default grid mode.

* Fixes/updates to the *BSD, Mac OS X, and MinGW builds.

* Changes in the gschem UI event state machine:

  - New drawing modes apply immediately.

  - Indication of an action being in progress has been added.


utils/gxyrs/lepton-xyrs.pl
#!/usr/bin/env perl
#
# Copyright (C) 2008 Carlos Nieves Onega
# Copyright (C) 2008 other contributors
#                        (see ChangeLog or SCM history for details)
 
# This file is part of lepton-xyrs.

# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  
# 02110-1301, USA.

use strict;
use warnings;

use lib 'GEDADATADIR/perl/lib'; # Where xyrs package is located

use xyrs;
use xyrs qw($CASE_INSENSITIVE);

# for parsing input options
use Getopt::Long;

use vars qw {
	$REF_COL $FOOTPRINT_COL $X_COL $Y_COL $ANGLE_COL
        $LAYER_COL $VALUE_COL $CASE_INSENSITIVE $LINE_NUMBER
        $LINE $TITLE_LINE
};

my $tempfile;
		

# Set these for debugging purposes
my $DEBUG = 0;
my $DEBUG_GUESSING = 0;
my $DEBUG_COL_LENGTH = 0;
my $DEBUG_RETURN_CODE = 0;

# Initialize global variables
$REF_COL = -1;
$FOOTPRINT_COL = -1;
$X_COL = -1;
$Y_COL = -1;
$ANGLE_COL = -1;
$LAYER_COL = -1;
$VALUE_COL = -1;

#######################################################################
#
# Subroutines
#
#######################################################################


#---------------------------------
# guess_file_format()
#
# Check current line and try to guess column numbers by 
# testing if it is the title line and there are some title keywords.
#---------------------------------
sub guess_file_format () {
    my $num_elems = scalar(@LINE);
    my @positions = ();
    my $position_repeated=0;

    if ($DEBUG_GUESSING == 1) {
	print "guess_file_format processing line $LINE_NUMBER.\n";
	print "Line: @LINE\n";
	print "Number of elements in line: $num_elems.\n";
    }

    # Remove '#' and spacer characters from the beginning of first element
    while ($LINE[0] =~ /^#/) {
	$LINE[0] = substr($LINE[0], 1);
    }
    while ($LINE[0] =~ /^\s+/) {
	$LINE[0] = substr($LINE[0], 1);
    }
    

    # Now try to guess column numbers from column titles.
    for (my $i = 0; $i <= $num_elems-1; $i++) {
	if ($DEBUG_GUESSING == 1) {
	    print "Testing element $i: ".$LINE[$i].".\n";
	}
	if ( ($LINE[$i] =~ /^Designator$/) ||
	     ($LINE[$i] =~ /^RefDesignator$/) ||
	     ($LINE[$i] =~ /^RefDes$/)
	    ) {
	    if ($DEBUG_GUESSING) {
		print "Found reference column: ".($i+1).".\n";
	    }
	    $REF_COL = $i+1;
	    if (grep(/\b$REF_COL\b/, @positions)) {
		$position_repeated = 1;
	    }
	    push(@positions,$REF_COL);
	}
	if ( ($LINE[$i] =~ /^Footprint$/) ||
	     ($LINE[$i] =~ /^TopCell$/) ||
	     ( ($LINE[$i] =~ /^Description$/) && 
	       (! grep(/\bTopCell\b/,@LINE)) )
	    ) {
	    if ($DEBUG_GUESSING) {
		print "Found footprint column: ".($i+1).".\n";
	    }
	    $FOOTPRINT_COL = $i+1;
	    if (grep(/\b$FOOTPRINT_COL\b/, @positions)) {
		$position_repeated = 1;
	    }
	    push(@positions,$FOOTPRINT_COL);
	}
	if ( ($LINE[$i] =~ /^Mid X$/) ||
	     ($LINE[$i] =~ /^X$/)
	    ) {
	    if ($DEBUG_GUESSING) {
		print "Found X column: ".($i+1).".\n";
	    }
	    $X_COL = $i+1;
	}
	if ( ($LINE[$i] =~ /^Mid Y$/) ||
	     ($LINE[$i] =~ /^Y$/)
	    ) {
	    if ($DEBUG_GUESSING) {
		print "Found Y column: ".($i+1).".\n";
	    }
	    $Y_COL = $i+1;
	    if (grep(/\b$Y_COL\b/, @positions)) {
		$position_repeated = 1;
	    }
	    push(@positions,$Y_COL);
	}
	if ( ($LINE[$i] =~ /^[Rr]otation$/) ||
	     ($LINE[$i] =~ /^Rot$/) 
	     
	    ) {
	    if ($DEBUG_GUESSING) {
		print "Found angle column: ".($i+1).".\n";
	    }
	    $ANGLE_COL = $i+1;
	    if (grep(/\b$ANGLE_COL\b/, @positions)) {
		$position_repeated = 1;
	    }
	    push(@positions,$ANGLE_COL);
	}
	if ( ($LINE[$i] =~ /^TB$/) ||
	     ($LINE[$i] =~ /^Side$/) ||
	     ($LINE[$i] =~ /^top\/bottom$/)
	    ) {
	    if ($DEBUG_GUESSING) {
		print "Found layer column: ".($i+1).".\n";
	    }
	    $LAYER_COL = $i+1;
	    if (grep(/\b$LAYER_COL\b/, @positions)) {
		$position_repeated = 1;
	    }
	    push(@positions,$LAYER_COL);
	}
	if ( ($LINE[$i] =~ /^Comment$/) ||
	     ($LINE[$i] =~ /^PartNumber$/) ||
	     ($LINE[$i] =~ /^Value$/)
	    ) {
	    if ($DEBUG_GUESSING) {
		print "Found value column: ".($i+1).".\n";
	    }
	    $VALUE_COL = $i+1; 	
	    if (grep(/\b$VALUE_COL\b/, @positions)) {
		$position_repeated = 1;
	    }
	    push(@positions,$VALUE_COL);
	}
    }

    # If there is any repeated position, warn the user.
    if ($position_repeated == 1) {
	print STDERR "A position was repeated while guessing file format.\n";
    }
    # If not all columns were guessed, or a position is repeated then reset all.
    if ( ($position_repeated == 1) ||
	 ($REF_COL == -1) ||
	 ($FOOTPRINT_COL == -1) ||
	 ($X_COL == -1) ||
	 ($Y_COL == -1) ||
	 ($ANGLE_COL == -1) ||
	 ($LAYER_COL == -1) ||
	 ($VALUE_COL == -1)
	) {

	# All columns should be defined in the same line.
	$REF_COL = -1;
	$FOOTPRINT_COL = -1;
	$X_COL = -1;
	$Y_COL = -1;
	$ANGLE_COL = -1;
	$LAYER_COL = -1;
	$VALUE_COL = -1;
    }

    if ($DEBUG_GUESSING == 1) {
	print "guess_file_format ended processing line $LINE_NUMBER.\n";
    }

}

#---------------------------------
# stdin()
#
# open a temp file, to allow multiple passes to the input file.
#---------------------------------
sub stdin () {
        $tempfile = 'temp_'.time() . '_' . int(rand(1000000));
        open(TEMP, "+>$tempfile") or print STDERR $!;
        local $/ = undef;
        print TEMP <STDIN>;
        seek(TEMP, 0, 0);
        \*TEMP;
}


#---------------------------------
# usage()
#
# prints program usage
#---------------------------------
sub usage {
    print "Usage:\n";
    print "  lepton-xyrs [--tabulate] <input_file> --adjust <adjust_file> --output <outputfile> \n";
    exit(0);
}


#######################################################################
#
# Main program
#
#######################################################################


{

    my @fname=();
    my ($file_in, $adjust_file, $file_out);
    my ($this_line, $line_out);
    my @lengths;
    my $output_delimiter="";

    # Delimiters to try when guessing file format.
    my @delimiters = ('\s+', ',\s*'); 
    my @delimiters_char = (' ', ','); # Used when writing the output file
    my $delimiter_index=-1; 
    
    my $pass;
    my $total_passes = scalar(@delimiters) + 1;
    my $guessing_file_format;
    my $is_comment = 0; # Mark the current line as a comment.
    my $eval_string;

    my $tabulate=0; # Output should be tabulated or not.
    my $verbose;
    my $process_comments;

    my $return_code;

    # Comparisons are case insensitive by default.
    $CASE_INSENSITIVE=0;

    &GetOptions(("help" => \&usage, 
		 "verbose" => \$verbose,
		 "process-comments" => \$process_comments,
		 "tabulate!" => \$tabulate,
		 "caseinsensitive!" => \$CASE_INSENSITIVE,
		 "adjust=s" => \$adjust_file,
		 "eval=s" => \$eval_string,
		 "output=s" => \$file_out,
		 "output-delimiter=s" => \$output_delimiter,
		));

    if ($Getopt::Long::error) {
	print STDERR "Wrong arguments.\n";
	usage();
    }
    
    # Print usage string in event user didn't call out any args
    usage() unless @ARGV;
    
    # Make sure the input schematic exists and we can read it.
    # Also count number of files to open.
    while(@ARGV) {
	my $i=0;
	$fname[$i]=shift(@ARGV);
	print "Checking argument $fname[$i].\n";
 	if (! -r $fname[$i]) {  
 	    die("Input file $fname[$i] does not exist or can not be read");
 	}
	$i++;
    }
    
    $file_in = $fname[0];
	
    # Print usage string in event user didn't specify input, adjust 
    # and output files.
    if (! ($file_in && $file_out && ($adjust_file || $eval_string))) {
	print STDERR "Nothing to do.\n";
	if ($eval_string) {
	    print STDERR "Eval string: $eval_string.\n";
	}
	if ($adjust_file) {
	    print STDERR "Commands file: $adjust_file.\n";
	}
	usage();
    }
    

    if ($file_in !~ '^-$') {
	open(FILE_IN, $file_in) || die ("Can't open input file $file_in: $!");
    }
    else {
	*FILE_IN = &stdin();	
    }

    if (length($output_delimiter) > 1) {
	die ("Output delimiter $output_delimiter must be a character.");
    }

    # Don't know yet the file format.
    $guessing_file_format = 1;
    $TITLE_LINE = 0;

    # Last pass is to process the file. Others are for guessing the
    # file format with several delimiters (one for each element in @delimiters.
    for ($pass = 1; $pass <= $total_passes; $pass++) {
	if (seek(FILE_IN, 0, 0) != 1) {
	    print STDERR "Error when seek input file to 0\n";
	}
	if ($pass >= $total_passes) {
	    open(FILE_OUT, "> ".$file_out) || die ("Can't open output file $file_out: $!");
	}
	
	# Set the return code to 0
	$return_code = 0;
	if ($DEBUG_RETURN_CODE) {
	    print "Setting global return code to 0.\n";
	}

	# Parse file
	$LINE_NUMBER = 0;
	print "Pass number: $pass.\n" if ($DEBUG);
	while ( <FILE_IN> ) {
	    $this_line=$_;
	    $_ = $this_line;

	    # Remove end line characters.
	    chomp($this_line);

	    # For the last pass, use the delimiter guessed.
	    # For the others, get the delimiter from @delimiters array.
	    if ($pass < $total_passes) {
		my $this_delimiter = $delimiters[$pass-1];
		@LINE = split ($this_delimiter, $this_line);
	    } else {
		my $this_delimiter = $delimiters[$delimiter_index];
		@LINE = split ($this_delimiter, $this_line);
	    }
	    $LINE_NUMBER += 1;
	    next if /^@/;
	    next if /^q/;
	    
	    # Test if the line is a comment 
	    # (which shouldn't be processed).
	    if ( ($this_line =~ /^\s*#/) ||
		 ($this_line =~ /^\s*$/) ) {
		if ($DEBUG) {
		    print "Line is comment: @LINE\n";
		}
		$is_comment = 1;
	    } else {
		$is_comment = 0;
	    }
	    
	    # First, some column titles are separated by spaces, so join them.
	    if ($LINE_NUMBER == 1) {
		my $num_elems = scalar(@LINE);
		for (my $i = 0; $i <= $num_elems-1; $i++) {
		    if ($i <= $num_elems-2) {
			# Join titles like "Pad X", "Pad Y", "Mid X", "Mid Y",
			# "ref X", "ref Y" (Protel output)
			if ( ( ($LINE[$i] =~ /^Pad$/) &&
			       ($LINE[$i+1] =~/^[XY]$/) ) ||
			     ( ($LINE[$i] =~ /^Mid$/) &&
			       ($LINE[$i+1] =~/^[XY]$/) ) ||
			     ( ($LINE[$i] =~ /^Ref$/) &&
			       ($LINE[$i+1] =~/^[XY]$/) ) 
			    ) {
			    $LINE[$i] = $LINE[$i]." ".$LINE[$i+1];
			    for (my $j = $i+1; $j <= $num_elems-2; $j++) {
				$LINE[$j] = $LINE[$j+1];
			    }
			    delete $LINE[$num_elems-1];
			    $num_elems -= 1;
			}
		    }
		}
		
	    }
	    
	    if ( ($pass != $total_passes) &&
		 $guessing_file_format ) {
		my $old_guessing_file_format = $guessing_file_format;
		# Try to guess the file format
		guess_file_format();
		# Check if already guessed the file format.
		$guessing_file_format = ( ($REF_COL == -1) ||
					  ($FOOTPRINT_COL == -1) ||
					  ($X_COL == -1) ||
					  ($Y_COL == -1) ||
					  ($ANGLE_COL == -1) ||
					  ($LAYER_COL == -1) ||
					  ($VALUE_COL == -1) );
		
		if (!$guessing_file_format) {
		    if ($old_guessing_file_format == 1) {
			$TITLE_LINE = $LINE_NUMBER;
		    }
		    if ($verbose) {
			print "Found file format. ";
			print "Delimiter is $delimiters[$pass-1].\n";
		    }
		    
		    $delimiter_index = $pass-1;
		}
	    }
	    
	    if (!$guessing_file_format) {
		if (($is_comment == 0) || $process_comments) {
		    my $rc ;

		    # If it's parsing the file then keep the number
		    # of columns constant, joining all the fields 
		    # in the last one.
 		    while ( (@LINE > @lengths) &&
			    (scalar(@lengths) > 0) ) {
 			$LINE[@LINE-2] = $LINE[@LINE-2].
			    $delimiters_char[$delimiter_index].$LINE[@LINE-1];
 			pop @LINE;
 		    }

		    if ($DEBUG_RETURN_CODE) {
			print "Global return code before evaluation: $return_code.\n";
		    }

		    # Run the adjust file.
		    if (defined $adjust_file) {
			$rc = do $adjust_file || 
			    die("Can't use adjust file: $!\n");
		    } else {
			#print "Eval: $eval_string\n";
			$rc = eval $eval_string;
		    }
		    if ($@) {
			print STDERR "Error: ".$@."\n";
			$rc = -10;
		    }
		    
		    # Set $return_code based on priorities.
		    # 1 means the command found a match and was succesful 
		    #   at least one time in the whole file.
		    # 0 means there was no match.
		    # -1 means there was an error.
		    # -2 means there was a warning.
		    if ($DEBUG_RETURN_CODE) {
			print "Return code $rc after evaluation.\n";
		    }
		    if ($rc == -1) {
			$return_code = -1;
		    }
		    elsif (($rc == -2) && ($return_code >= 0)) {
			$return_code = -2;
		    }
		    elsif (($rc == 1) && ($return_code == 0)) {
			$return_code = 1;
		    }
		    elsif (($rc == 0) && ($return_code == 1)) {
			$return_code = 1;
		    }
		    elsif (($rc >= 0) && ($return_code < 0)) {
			# Do nothing
		    }
		    else {
			$return_code = $rc;
		    }
		    if ($DEBUG_RETURN_CODE) {
			print "Global return code: $return_code.\n";
		    }
		}	
		if ( ($pass != $total_passes) &&
		     ( ($is_comment == 0) || $process_comments) ) {
		    if ($DEBUG_COL_LENGTH == 1) {
			print "Measuring column length of line: ".
			    "'$this_line'\n"; 
		    }
		    # Calcule max length array
		    if ($tabulate) {
			my $j=0;
			for (my $i=0; $i <= scalar(@LINE)-1; $i++) {
			    # Allow adding more lines, and checking
			    # element's length right.
			    if ($LINE[$i] eq "\n") {
				$j = $i+1;
				next;
			    }
			    if (! exists $lengths[$i-$j]) {
				$lengths[$i-$j]= length($LINE[$i]);
			    }
			    elsif ($lengths[$i-$j] < length($LINE[$i]) ) {
				$lengths[$i-$j]= length($LINE[$i]);
			    }
			    if ($DEBUG_COL_LENGTH == 1) {
				print "Column ".($i-$j+1).
				    ", length: $lengths[$i].\n"; 
			    }
			}
		    }
		}
		
		# Print the result after processing
		if ($pass >= $total_passes) {
		    if (($is_comment == 1) && !$process_comments) {
			print FILE_OUT "$this_line\n";
		    }
		    elsif (@LINE > 0) {
			my $delimiter;
			my $j=0;
			
			# Set the output delimiter 
			if (length($output_delimiter) != 0) {
			    $delimiter = $output_delimiter;
			}
			else {
			    $delimiter = $delimiters_char[$delimiter_index];
			}

			# Write output
			for (my $i=0; $i <= scalar(@LINE)-1; $i++) {
			    # Allow adding more lines, and handle
			    # new lines.
			    if ($LINE[$i] eq "\n") {
				print FILE_OUT "\n";
				$j = $i+1;
				next;
			    }
			    if (($i-$j) > scalar(@lengths)-1) {
				# If column length array has no number 
				# for this column, print it as is.
				print FILE_OUT $LINE[$i];
				} else {
				    printf FILE_OUT "%-$lengths[$i-$j]s",$LINE[$i];
				}
			    # print the last column without delimiter
			    if ( ($i < @LINE-1) &&
				 (!($LINE[$i+1] eq "\n"))) {
				    print FILE_OUT $delimiter;
			    }
			}
			print FILE_OUT "\n";
		    }
		}
	    }
	    
	}	
	# Close output file.
	if ($pass >= $total_passes) {
	    close(FILE_OUT);
	}
	
	# Print all column numbers (DEBUGGING)
	if ( ($pass != $total_passes) && ($DEBUG_GUESSING==1) ) {
	    print "Reference column: $REF_COL.\n";
	    print "Footprint column: $FOOTPRINT_COL.\n";
	    print "X column:         $X_COL.\n";
	    print "Y column:         $Y_COL.\n";
	    print "Rotation column:  $ANGLE_COL.\n";
	    print "Layer column:     $LAYER_COL.\n";
	    print "Value column:     $VALUE_COL.\n";
	}

	# If last guessing pass is complete, but all columns were not guessed, 
	# exit with an error.
	if ( $guessing_file_format &&
	     ($pass == $total_passes-1) )
	{
	    die ("Can't guess file format.\n");
	}

	# If file format was guessed, but next pass is not the last one,
	# then go to the last one (processing pass).
	if ( (!$guessing_file_format) &&
	     ($pass < $total_passes-1) ) {
	    $pass = $total_passes-1;
	}
    }        
    close(FILE_IN);
    if ($tempfile) {
	unlink $tempfile;
    }

    exit $return_code;
}

